File: core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.java
package fr.sncf.osrd.api.pathfinding;

import static fr.sncf.osrd.api.pathfinding.RemainingDistanceEstimator.minDistanceBetweenSteps;
import static fr.sncf.osrd.railjson.schema.common.graph.EdgeDirection.START_TO_STOP;
import static fr.sncf.osrd.railjson.schema.common.graph.EdgeDirection.STOP_TO_START;
import static fr.sncf.osrd.sim_infra.api.TrackInfraKt.getTrackSectionFromNameOrThrow;
import static fr.sncf.osrd.sim_infra.utils.InfraUtilsKt.getNextTrackSections;
import static fr.sncf.osrd.utils.KtToJavaConverter.toIntList;
import static fr.sncf.osrd.utils.KtToJavaConverter.toIntSet;
import static fr.sncf.osrd.utils.indexing.DirStaticIdxKt.toDirection;
import static fr.sncf.osrd.utils.indexing.DirStaticIdxKt.toValue;
import static fr.sncf.osrd.utils.units.Distance.fromMeters;
import static fr.sncf.osrd.utils.units.Distance.toMeters;

import fr.sncf.osrd.api.ExceptionHandler;
import fr.sncf.osrd.api.FullInfra;
import fr.sncf.osrd.api.InfraManager;
import fr.sncf.osrd.api.pathfinding.constraints.ElectrificationConstraints;
import fr.sncf.osrd.api.pathfinding.constraints.LoadingGaugeConstraints;
import fr.sncf.osrd.api.pathfinding.request.PathfindingRequest;
import fr.sncf.osrd.api.pathfinding.request.PathfindingWaypoint;
import fr.sncf.osrd.api.pathfinding.response.PathfindingResult;
import fr.sncf.osrd.infra.api.Direction;
import fr.sncf.osrd.railjson.parser.RJSRollingStockParser;
import fr.sncf.osrd.railjson.schema.common.graph.EdgeDirection;
import fr.sncf.osrd.railjson.schema.infra.trackranges.RJSDirectionalTrackRange;
import fr.sncf.osrd.reporting.exceptions.ErrorType;
import fr.sncf.osrd.reporting.exceptions.OSRDError;
import fr.sncf.osrd.reporting.warnings.DiagnosticRecorderImpl;
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra;
import fr.sncf.osrd.train.RollingStock;
import fr.sncf.osrd.utils.graph.GraphAdapter;
import fr.sncf.osrd.utils.graph.Pathfinding;
import fr.sncf.osrd.utils.graph.functional_interfaces.AStarHeuristic;
import fr.sncf.osrd.utils.graph.functional_interfaces.EdgeToRanges;
import fr.sncf.osrd.utils.indexing.DirStaticIdxKt;
import org.takes.Request;
import org.takes.Response;
import org.takes.Take;
import org.takes.rq.RqPrint;
import org.takes.rs.RsJson;
import org.takes.rs.RsText;
import org.takes.rs.RsWithBody;
import org.takes.rs.RsWithStatus;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

public class PathfindingBlocksEndpoint implements Take {
    private final InfraManager infraManager;
    private static final HashMap<Class<?>, ErrorType> constraintErrors = new HashMap<>();

    static {
        constraintErrors.put(LoadingGaugeConstraints.class, ErrorType.PathfindingGaugeError);
        constraintErrors.put(ElectrificationConstraints.class, ErrorType.PathfindingElectrificationError);
    }

    /**
     * constructor
     */
    public PathfindingBlocksEndpoint(InfraManager infraManager) {
        this.infraManager = infraManager;
    }

    @Override
    public Response act(Request req) {
        var recorder = new DiagnosticRecorderImpl(false);
        try {
            var body = new RqPrint(req).printBody();
            var request = PathfindingRequest.adapter.fromJson(body);
            if (request == null)
                return new RsWithStatus(new RsText("missing request body"), 400);

            var reqWaypoints = request.waypoints;

            // Load infra
            var infra = infraManager.getInfra(request.infra, request.expectedVersion, recorder);

            // Load rolling stocks
            var rollingStocks = List.<RollingStock>of();
            if (request.rollingStocks != null)
                rollingStocks = request.rollingStocks.stream()
                        .map(RJSRollingStockParser::parse)
                        .toList();

            var path = runPathfinding(infra, reqWaypoints, rollingStocks);

            PathfindingResult res = PathfindingResultConverter.convert(infra.blockInfra(), infra.rawInfra(),
                    path, recorder);

            validatePathfindingResult(res, reqWaypoints, infra.rawInfra());

            return new RsJson(new RsWithBody(PathfindingResult.adapterResult.toJson(res)));
        } catch (Throwable ex) {
            // TODO: include warnings in the response
            return ExceptionHandler.handle(ex);
        }
    }

    /**
     * Runs the pathfinding with the infra and rolling stocks already parsed
     */
    public static Pathfinding.Result<Integer> runPathfinding(
            FullInfra infra,
            PathfindingWaypoint[][] reqWaypoints,
            Collection<RollingStock> rollingStocks
    ) throws OSRDError {
        // Parse the waypoints
        var waypoints = new ArrayList<Collection<Pathfinding.EdgeLocation<Integer>>>();
        for (var step : reqWaypoints) {
            var allStarts = new HashSet<Pathfinding.EdgeLocation<Integer>>();
            for (var waypoint : step)
                allStarts.addAll(findWaypointBlocks(infra, waypoint));
            waypoints.add(allStarts);
        }

        // Initializes the constraints
        var loadingGaugeConstraints = new LoadingGaugeConstraints(infra.blockInfra(), infra.rawInfra(), rollingStocks);
        var electrificationConstraints = new ElectrificationConstraints(infra.blockInfra(), infra.rawInfra(),
                rollingStocks);
        final List<EdgeToRanges<Integer>> constraints = List.of(loadingGaugeConstraints, electrificationConstraints);
        var remainingDistanceEstimators = makeHeuristics(infra, waypoints);

        // Compute the paths from the entry waypoint to the exit waypoint
        return computePaths(infra, waypoints, constraints, remainingDistanceEstimators);
    }

    /** Initialize the heuristics */
    public static ArrayList<AStarHeuristic<Integer>> makeHeuristics(
            FullInfra infra,
            List<Collection<Pathfinding.EdgeLocation<Integer>>> waypoints
    ) {
        // Compute the minimum distance between steps
        double[] stepMinDistance = new double[waypoints.size() - 1];
        for (int i = 0; i < waypoints.size() - 2; i++) {
            stepMinDistance[i] = minDistanceBetweenSteps(infra.blockInfra(), infra.rawInfra(), waypoints.get(i + 1),
                    waypoints.get(i + 2));
        }

        // Reversed cumulative sum
        for (int i = stepMinDistance.length - 2; i >= 0; i--) {
            stepMinDistance[i] += stepMinDistance[i + 1];
        }

        // Setup estimators foreach intermediate steps
        var remainingDistanceEstimators = new ArrayList<AStarHeuristic<Integer>>();
        for (int i = 0; i < waypoints.size() - 1; i++) {
            remainingDistanceEstimators.add(new RemainingDistanceEstimator(
                    infra.blockInfra(),
                    infra.rawInfra().getSimInfra(),
                    waypoints.get(i + 1),
                    stepMinDistance[i]
            ));
        }
        return remainingDistanceEstimators;
    }

    private static Pathfinding.Result<Integer> computePaths(
            FullInfra infra,
            ArrayList<Collection<Pathfinding.EdgeLocation<Integer>>> waypoints,
            List<EdgeToRanges<Integer>> constraints,
            List<AStarHeuristic<Integer>> remainingDistanceEstimators
    ) throws OSRDError {
        var pathFound = new Pathfinding<>(new GraphAdapter(infra.blockInfra(), infra.rawInfra()))
                .setEdgeToLength(block -> infra.blockInfra().getBlockLength(block))
                .setRemainingDistanceEstimator(remainingDistanceEstimators)
                .addBlockedRangeOnEdges(constraints)
                .runPathfinding(waypoints);

        if (pathFound != null) {
            return pathFound;
        }

        // Handling errors
        // Check if pathfinding failed due to constraints
        var possiblePathWithoutErrorNoConstraints =
                new Pathfinding<>(new GraphAdapter(infra.blockInfra(), infra.rawInfra()))
                        .setEdgeToLength(block -> infra.blockInfra().getBlockLength(block))
                        .setRemainingDistanceEstimator(remainingDistanceEstimators)
                        .runPathfinding(waypoints);
        if (possiblePathWithoutErrorNoConstraints != null) {
            for (EdgeToRanges<Integer> currentConstraint : constraints) {
                var possiblePathWithoutError = new Pathfinding<>(new GraphAdapter(infra.blockInfra(), infra.rawInfra()))
                        .setEdgeToLength(block -> infra.blockInfra().getBlockLength(block))
                        .addBlockedRangeOnEdges(currentConstraint)
                        .setRemainingDistanceEstimator(remainingDistanceEstimators)
                        .runPathfinding(waypoints);
                if (possiblePathWithoutError == null) {
                    throw new OSRDError(constraintErrors.get(currentConstraint.getClass()));
                }
            }
        }
        // It didn’t fail due to a constraint, no path exists
        throw new OSRDError(ErrorType.PathfindingGenericError);
    }

    /**
     * Returns all the EdgeLocations of a waypoint list.
     * @param infra full infra.
     * @param waypoints corresponding waypoints.
     * @return corresponding edge locations.
     */
    public static Set<Pathfinding.EdgeLocation<Integer>> findWaypointBlocks(FullInfra infra,
                                                                            Collection<PathfindingWaypoint> waypoints) {
        return waypoints.stream()
                .flatMap(waypoint -> findWaypointBlocks(infra, waypoint).stream())
                .collect(Collectors.toSet());
    }

    /**
     * Returns all the EdgeLocations of a waypoint.
     * @param infra full infra.
     * @param waypoint corresponding waypoint.
     * @return corresponding edge location, containing a block id and its offset from the waypoint.
     */
    public static Set<Pathfinding.EdgeLocation<Integer>> findWaypointBlocks(
            FullInfra infra,
            PathfindingWaypoint waypoint
    ) {
        var res = new HashSet<Pathfinding.EdgeLocation<Integer>>();
        var trackSectionId = getTrackSectionFromNameOrThrow(waypoint.trackSection, infra.rawInfra());
        var trackChunkOnWaypoint = getTrackSectionChunkOnWaypoint(trackSectionId, waypoint.offset, infra.rawInfra());
        var waypointDirection = Direction.fromEdgeDir(waypoint.direction).toKtDirection();
        var blocksOnWaypoint =
                toIntSet(infra.blockInfra().getBlocksFromTrackChunk(trackChunkOnWaypoint, waypointDirection));
        blocksOnWaypoint.forEach(block -> {
            var offset = getBlockOffset(block, trackChunkOnWaypoint, trackSectionId, waypoint.offset,
                    waypoint.direction, infra);
            res.add(new Pathfinding.EdgeLocation<>(block, offset));
        });
        return res;
    }

    private static Integer getTrackSectionChunkOnWaypoint(int trackSectionId, double waypointOffset,
                                                          RawSignalingInfra rawInfra) {
        var waypointOffsetMilli = fromMeters(waypointOffset);
        var trackSectionChunks = toIntList(rawInfra.getTrackSectionChunks(trackSectionId));
        return trackSectionChunks.stream()
                .filter(chunk -> {
                    var startChunk = rawInfra.getTrackChunkOffset(chunk);
                    var endChunk = startChunk + rawInfra.getTrackChunkLength(chunk);
                    return waypointOffsetMilli >= startChunk && waypointOffsetMilli <= endChunk;
                })
                .findFirst()
                .orElseThrow(() -> new RuntimeException(
                        String.format("The waypoint is not located on the track section %s", trackSectionId)));

    }

    private static double getBlockOffset(int blockId, int trackChunkId, int trackSectionId, double waypointOffsetMeters,
                                         EdgeDirection direction, FullInfra infra) {
        var waypointOffset = fromMeters(waypointOffsetMeters);
        var trackSectionLength = infra.rawInfra().getTrackSectionLength(trackSectionId);
        var trackChunkOffset = infra.rawInfra().getTrackChunkOffset(trackChunkId);
        var trackChunkLength = infra.rawInfra().getTrackChunkLength(trackChunkId);
        var dirTrackChunkOffset = direction.equals(START_TO_STOP)
                ? trackChunkOffset
                : trackSectionLength - trackChunkOffset - trackChunkLength;
        var dirWaypointOffset = direction.equals(START_TO_STOP)
                ? waypointOffset
                : trackSectionLength - waypointOffset;
        var startBlockToStartChunk = 0L;
        var blockTrackChunks = toIntList(infra.blockInfra().getTrackChunksFromBlock(blockId));
        for (Integer blockTrackChunkDirId: blockTrackChunks) {
            var blockTrackChunkId = toValue(blockTrackChunkDirId);
            if (blockTrackChunkId == trackChunkId) {
                return Math.abs(startBlockToStartChunk + dirWaypointOffset - dirTrackChunkOffset);
            }
            startBlockToStartChunk += infra.rawInfra().getTrackChunkLength(blockTrackChunkId);
        }
        throw new AssertionError(
                String.format("getBlockOffset: Track chunk %s not in block %s", trackChunkId, blockId));
    }

    /** Returns the list of dir track id on the route. Can contain duplicates. */
    private List<Integer> getRouteDirTracks(RawSignalingInfra rawInfra, int routeId) {
        var res = new ArrayList<Integer>();
        for (var dirChunk : toIntList(rawInfra.getChunksOnRoute(routeId))) {
            var currentTrack = DirStaticIdxKt.from(
                    rawInfra.getTrackFromChunk(toValue(dirChunk)),
                    toDirection(dirChunk)
            );
            res.add(currentTrack);
        }
        return res;
    }

    private void validatePathfindingResult(PathfindingResult res, PathfindingWaypoint[][] reqWaypoints,
                                           RawSignalingInfra rawInfra) {
        var routeTracks = res.routePaths.stream()
                .flatMap(rjsRoutePath ->
                        getRouteDirTracks(rawInfra, rawInfra.getRouteFromName(rjsRoutePath.route)).stream())
                .distinct()
                .toList();
        assertPathRoutesAreAdjacent(routeTracks, rawInfra);

        var tracksOnPath = res.routePaths.stream()
                .flatMap(route -> route.trackSections.stream())
                .toList();
        assertPathTracksAreComplete(tracksOnPath, rawInfra);
        assertRequiredWaypointsOnPathTracks(reqWaypoints, tracksOnPath);
    }

    private void assertPathRoutesAreAdjacent(List<Integer> routeTracks, RawSignalingInfra rawInfra) {
        for (int i = 0; i < routeTracks.size() - 1; i++) {
            var nextTrackSections = toIntList(getNextTrackSections(rawInfra, routeTracks.get(i)));
            assert nextTrackSections.contains(routeTracks.get(i + 1))
                    : "The path goes over consecutive tracks that are not adjacent";
        }
    }

    private void assertPathTracksAreComplete(List<RJSDirectionalTrackRange> tracksOnPath, RawSignalingInfra rawInfra) {
        int i = 0;
        while (i < tracksOnPath.size()) {
            var track = tracksOnPath.get(i);
            var trackName = track.trackSectionID;
            var fullTrack = new ArrayList<>(tracksOnPath.stream()
                    .filter(trackOnPath -> Objects.equals(trackOnPath.trackSectionID, track.trackSectionID))
                    .toList());
            var fullTrackLength =
                    toMeters(rawInfra.getTrackSectionLength(getTrackSectionFromNameOrThrow(trackName, rawInfra)));

            var beginEndpoint = fullTrack.get(0).begin;
            var endEndpoint = fullTrack.get(fullTrack.size() - 1).end;
            var trackBegin = 0.;
            var trackEnd = fullTrackLength;
            if (track.direction.equals(STOP_TO_START)) {
                beginEndpoint = fullTrack.get(0).end;
                endEndpoint = fullTrack.get(fullTrack.size() - 1).begin;
                trackBegin = fullTrackLength;
                trackEnd = 0.;
                Collections.reverse(fullTrack);
            }

            // The first track starts at the first waypoint offset, hence it does not need to start at a track endpoint
            assert i == 0 || beginEndpoint == trackBegin
                    : String.format("The path goes through the track %s without starting at its endpoint", trackName);

            i += fullTrack.size();
            // The last track ends at the last waypoint offset, hence it does not need to end at a track endpoint
            assert i == tracksOnPath.size() || Math.abs(endEndpoint - trackEnd) < 1e-3
                    : String.format("The path goes through the track %s without ending at its endpoint", trackName);

            // The track should be whole
            for (int j = 0; j < fullTrack.size() - 1; j++) {
                assert Math.abs(fullTrack.get(j).end - fullTrack.get(j + 1).begin) < 1e-3
                        : String.format("The path goes through a partially incomplete track %s", trackName);
            }
        }
    }

    private void assertRequiredWaypointsOnPathTracks(PathfindingWaypoint[][] reqWaypoints,
                                                     List<RJSDirectionalTrackRange> tracksOnPath) {
        // Checks that at least one waypoint of each step is on the path
        assert Arrays.stream(reqWaypoints).allMatch(step -> Arrays.stream(step)
                .anyMatch(waypoint -> tracksOnPath.stream()
                        .anyMatch(trackOnPath -> isWaypointOnTrack(waypoint, trackOnPath))))
                : "The path does not contain one of the wanted steps";
    }

    private boolean isWaypointOnTrack(PathfindingWaypoint waypoint, RJSDirectionalTrackRange track) {
        return Objects.equals(track.trackSectionID, waypoint.trackSection)
                && Objects.equals(track.direction, waypoint.direction)
                && (track.begin <= waypoint.offset || Math.abs(track.begin - waypoint.offset) < 1e-3)
                && (track.end >= waypoint.offset || Math.abs(track.end - waypoint.offset) < 1e-3);
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt
package fr.sncf.osrd.api.pathfinding

import fr.sncf.osrd.api.ExceptionHandler
import fr.sncf.osrd.api.FullInfra
import fr.sncf.osrd.api.InfraProvider
import fr.sncf.osrd.api.TrackLocation
import fr.sncf.osrd.api.stdcm.parseTrackSectionIds
import fr.sncf.osrd.cli.Request
import fr.sncf.osrd.cli.Response
import fr.sncf.osrd.cli.RsJson
import fr.sncf.osrd.cli.RsText
import fr.sncf.osrd.cli.RsWithBody
import fr.sncf.osrd.cli.RsWithStatus
import fr.sncf.osrd.cli.Take
import fr.sncf.osrd.graph.*
import fr.sncf.osrd.path.interfaces.PhysicsPath
import fr.sncf.osrd.path.interfaces.TrainPath
import fr.sncf.osrd.pathfinding.Pathfinding
import fr.sncf.osrd.pathfinding.constraints.initConstraintsFromRSProps
import fr.sncf.osrd.reporting.exceptions.ErrorType
import fr.sncf.osrd.reporting.exceptions.OSRDError
import fr.sncf.osrd.sim_infra.api.*
import fr.sncf.osrd.stdcm.infra_exploration.ExplorerStep
import fr.sncf.osrd.stdcm.infra_exploration.InfraExplorer
import fr.sncf.osrd.utils.*
import fr.sncf.osrd.utils.units.Length
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.meters
import fr.sncf.osrd.utils.units.toDirected
import io.opentelemetry.api.trace.Span
import io.opentelemetry.api.trace.SpanKind
import io.opentelemetry.instrumentation.annotations.WithSpan
import java.io.File
import java.time.Duration
import java.time.Instant
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * Exception used to wrap the response when we can't find a path. We do want to interrupt the
 * process so an exception is relevant, but it's local to this file because the response should be a
 * 200
 */
class NoPathFoundException(val response: PathfindingBlockResponse) : Exception()

val pathfindingLogger: Logger = LoggerFactory.getLogger("Pathfinding")

class PathfindingBlocksEndpoint(private val infraManager: InfraProvider) : Take {
    override fun act(req: Request, ctx: Take.QueueContext?): Response {
        val body = req.body()
        val request =
            pathfindingRequestAdapter.fromJson(body)
                ?: return RsWithStatus(RsText("Missing request body"), 400)

        val logRequest = System.getenv("LOG_PATHFINDING_REQUESTS")
        if (logRequest?.equals("true", ignoreCase = true) == true) {
            val time = LocalDateTime.now()
            val formatted = time.format(DateTimeFormatter.ofPattern("MM-dd-HH:mm:ss:SSS"))
            val filename = "pathfinding-$formatted.json"
            Span.current()?.setAttribute("request-file", filename)
            File(filename).printWriter().use {
                it.println(pathfindingRequestAdapter.indent("    ").toJson(request))
            }
        }

        return run(request)
    }

    @WithSpan(value = "Processing pathfinding request", kind = SpanKind.SERVER)
    fun run(request: PathfindingBlockRequest): Response {
        try {
            // Load infra
            val infra = infraManager.getInfra(request.infra, request.expectedVersion)
            val res = runPathfinding(infra, request)
            pathfindingLogger.info("Success")
            return RsJson(RsWithBody(pathfindingResponseAdapter.toJson(res)))
        } catch (error: NoPathFoundException) {
            pathfindingLogger.info("No path found")
            return RsJson(RsWithBody(pathfindingResponseAdapter.toJson(error.response)))
        } catch (ex: Throwable) {
            pathfindingLogger.info("Pathfinding failed: ${ex.message}")
            return ExceptionHandler.handle(ex)
        }
    }
}

/** Runs the pathfinding with the infra and request already parsed */
@Throws(OSRDError::class)
fun runPathfinding(infra: FullInfra, request: PathfindingBlockRequest): PathfindingBlockResponse {
    // Parse the waypoints
    val targets = ArrayList<ExplorerStep>()
    val destinationTrack = request.pathItems.last()
    val destinationBlock = findWaypointBlocks(infra, destinationTrack.locations)
    request.pathItems.forEachIndexed { stepIndex, step ->
        val allStarts = HashSet<BlockLocation>()
        for (direction in Direction.entries) {
            for (waypoint in step.locations) {
                val waypointBlocks = findDirectedWaypointBlocks(infra, waypoint, direction)
                if (
                    request.stopsAtEndOfBlock == true &&
                        stepIndex != 0 &&
                        stepIndex != request.pathItems.size - 1
                ) {
                    allStarts.addAll(
                        waypointBlocks.map {
                            findStopPositionAtEndOfBlockConsideringRollingStock(
                                it,
                                destinationBlock,
                                request.rollingStockLength,
                                infra,
                            )
                        }
                    )
                } else {
                    allStarts.addAll(waypointBlocks)
                }
            }
        }
        targets.add(ExplorerStep(allStarts, canBacktrack = step.canBacktrack))
    }
    if (targets.size < 2) throw NoPathFoundException(NotEnoughPathItems())
    val allowedTrackSectionIds = parseTrackSectionIds(infra, request.allowedTrackSections)
    val constraints =
        initConstraintsFromRSProps(
            infra,
            request.rollingStockIsThermal,
            request.rollingStockLoadingGauge,
            request.rollingStockSupportedElectrifications,
            request.rollingStockSupportedSignalingSystems,
            allowedTrackSectionIds,
        )

    // Compute the paths from the entry waypoint to the exit waypoint
    val timeout = request.timeout ?: Pathfinding.TIMEOUT
    val path = computePaths(infra, targets, constraints, request, timeout)
    return runPathfindingPostProcessing(infra, request, path)
}

@Throws(OSRDError::class)
private fun computePaths(
    infra: FullInfra,
    targets: ArrayList<ExplorerStep>,
    constraints: List<PathfindingConstraint>,
    initialRequest: PathfindingBlockRequest,
    timeout: Double,
): ProcessedPathfindingResponse {
    val start = Instant.now()
    val pathFound =
        Pathfinding(
                infra,
                targets,
                constraints,
                initialRequest.speedLimitTag,
                initialRequest.rollingStockMaximumSpeed,
                initialRequest.rollingStockLength.meters,
            )
            .runPathfinding()

    if (pathFound != null) {
        pathfindingLogger.info("Path found, start postprocessing")
        val res = processPathfindingResponse(infra, pathFound)
        if (!hasDuplicateTracks(infra, res.path)) return res
        else pathfindingLogger.info("Path has duplicate tracks, dismissing")
    }

    // Handling errors
    // Check if pathfinding failed due to incompatible constraints
    pathfindingLogger.info("No path found, identifying issues")
    val elapsedSeconds = Duration.between(start, Instant.now()).toSeconds()
    throw buildNoPathFoundException(
        infra,
        targets,
        constraints,
        initialRequest,
        timeout.minus(elapsedSeconds),
    )
}

/**
 * Return true if the path contains a duplicated track. This kind of path is not supported by OSRD
 * yet.
 */
fun hasDuplicateTracks(infra: FullInfra, path: TrainPath): Boolean {
    val tracks =
        path
            .getChunks()
            .map { it.value }
            .map { DirTrackSectionId(infra.rawInfra.getTrackFromChunk(it.value), it.direction) }
            .withoutConsecutiveDuplicates()
    return tracks.toSet().size < tracks.size
}

@WithSpan(value = "Identifying why no path was found")
private fun buildNoPathFoundException(
    infra: FullInfra,
    targets: ArrayList<ExplorerStep>,
    constraints: Collection<PathfindingConstraint>,
    initialRequest: PathfindingBlockRequest,
    timeout: Double,
): NoPathFoundException {
    try {
        val possiblePathWithoutErrorNoConstraints =
            Pathfinding(
                    infra,
                    targets,
                    listOf(),
                    initialRequest.speedLimitTag,
                    initialRequest.rollingStockMaximumSpeed,
                    initialRequest.rollingStockLength.meters,
                )
                .runPathfinding(timeout)
        if (possiblePathWithoutErrorNoConstraints != null) {
            buildIncompatibleConstraintsResponse(
                    infra,
                    processPathfindingResponse(infra, possiblePathWithoutErrorNoConstraints),
                    constraints,
                    initialRequest,
                )
                ?.let {
                    return NoPathFoundException(it)
                }
        }
    } catch (error: OSRDError) {
        if (error.osrdErrorType == ErrorType.PathfindingTimeoutError) {
            throw OSRDError(ErrorType.PathfindingRelaxedPathTimeoutError)
        }
        throw error
    }
    // It didn’t fail due to an incompatible constraint, no path exists
    return NoPathFoundException(NotFoundInBlocks(listOf(), Length(0.meters)))
}

data class ProcessedPathfindingResponse(
    val path: TrainPath,
    val waypointOffsets: List<Offset<PhysicsPath>>,
    val backtrackIndexes: List<Int>,
)

private fun processPathfindingResponse(
    infra: FullInfra,
    explorer: InfraExplorer,
): ProcessedPathfindingResponse {
    val trainPath = explorer.buildFullPath(infra.rawInfra, infra.blockInfra)
    val seenSteps = explorer.getStepTracker().getSeenSteps().toList()
    val stepOffsets = seenSteps.map { it.travelledPathOffset }
    val backtrackIndexes =
        seenSteps.mapIndexedNotNull { index, step -> if (step.isBacktracking) index else null }
    return ProcessedPathfindingResponse(trainPath, stepOffsets, backtrackIndexes)
}

/**
 * Returns all the EdgeLocations of a waypoint.
 *
 * @param infra full infra.
 * @param waypoint corresponding waypoint.
 * @return corresponding edge location, containing a block id and its offset from the waypoint.
 */
fun findDirectedWaypointBlocks(
    infra: FullInfra,
    waypoint: TrackLocation,
    direction: Direction,
): Set<BlockLocation> {
    val res = HashSet<BlockLocation>()
    val trackSectionId =
        infra.rawInfra.getTrackSectionFromName(waypoint.track)
            ?: throw OSRDError.newUnknownTrackSectionError(waypoint.track)

    val chunkLocationOnWaypoint =
        getChunkLocationOnWaypoint(trackSectionId, waypoint.offset, infra.rawInfra)
    val chunkLength = infra.rawInfra.getTrackChunkLength(chunkLocationOnWaypoint.chunk)
    val directedChunkOffset = chunkLocationOnWaypoint.offset.toDirected(chunkLength, direction)
    val waypointDirChunkLocation =
        DirChunkLocation(
            DirTrackChunkId(chunkLocationOnWaypoint.chunk, direction),
            directedChunkOffset,
        )

    val blocksOnWaypoint =
        infra.blockInfra.getBlocksFromTrackChunk(chunkLocationOnWaypoint.chunk, direction).toSet()
    for (block in blocksOnWaypoint) {
        val offset =
            infra.blockInfra.getBlockOffset(block, waypointDirChunkLocation, infra.rawInfra)
        assert(offset <= infra.blockInfra.getBlockLength(block))
        res.add(BlockLocation(block, offset))
    }
    return res
}

fun findWaypointBlocks(infra: FullInfra, waypoints: Collection<TrackLocation>): Set<BlockLocation> {
    val waypointBlocks = HashSet<BlockLocation>()
    for (waypoint in waypoints) {
        for (direction in Direction.entries) {
            waypointBlocks.addAll(findDirectedWaypointBlocks(infra, waypoint, direction))
        }
    }
    return waypointBlocks
}

private fun getChunkLocationOnWaypoint(
    trackSectionId: TrackSectionId,
    waypointOffset: Offset<TrackSection>,
    rawInfra: RawSignalingInfra,
): ChunkLocation {
    val trackSectionChunks = rawInfra.getTrackSectionChunks(trackSectionId)
    for (chunk in trackSectionChunks) {
        val startChunk = rawInfra.getTrackChunkOffset(chunk)
        val endChunk = startChunk + rawInfra.getTrackChunkLength(chunk).distance
        if (waypointOffset in startChunk..endChunk)
            return ChunkLocation(chunk, Offset(waypointOffset - startChunk))
    }
    throw OSRDError(ErrorType.InvalidWaypointLocation)
        .withContext("track", rawInfra.getTrackSectionName(trackSectionId))
        .withContext("offset", waypointOffset)
}

/**
 * Given a waypoint block (initial position of a train stop), returns the new stop position of a
 * train: end of current block, with the constraints of staying in the same block and keeping the
 * tail of the train on the initial position (waypointBlock). Note that nodes of the infra are not
 * considered.
 */
fun findStopPositionAtEndOfBlockConsideringRollingStock(
    waypointBlock: BlockLocation,
    destinationBlock: Set<BlockLocation>,
    rollingStockLength: Double,
    infra: FullInfra,
): BlockLocation {
    // To ensure that the tail of the rolling stock is not further than the initial operational
    // point (waypointBlock) position
    val maxTailOffset = waypointBlock.offset + rollingStockLength.meters

    // Can't go further than the end of the block, which is delimited by a block-delimiting signal
    val maxHeadOffset = infra.blockInfra.getBlockLength(waypointBlock.edge)

    val newWaypointOffset = Offset.min(maxTailOffset, maxHeadOffset)

    val destinationOffset =
        destinationBlock
            .filter { it.edge == waypointBlock.edge && waypointBlock.offset <= it.offset }
            .minOfOrNull { it.offset }

    // The waypoint offset does not change in the case where the stop is after the destination
    if (destinationOffset != null && destinationOffset <= newWaypointOffset)
        return BlockLocation(waypointBlock.edge, waypointBlock.offset)

    return BlockLocation(waypointBlock.edge, newWaypointOffset)
}
